golang/subscriber.go

```diff
@@ -41,135 +47,180 @@ func (m *message) GetEvent() *v2.Event {
 	ackStream       proxypb.StoreProxy_AckClient
-	messagec        chan Message
+	messageC        chan Message
 	state           streamState
 	mu              sync.Mutex
+	closeOnce       sync.Once
+	handler         func(ctx context.Context, msgs ...Message) error
+	closeC          chan struct{}
 }
 
-func newSubscribe(
-	c *client,
-	opts *SubscribeOptions,
-) (*subscribe, error) {
-	in := &proxypb.SubscribeRequest{
-		SubscriptionId: opts.SubscriptionID,
-	}
-	subscribeStream, err := c.store.Subscribe(context.Background(), in)
-	if err != nil {
-		return nil, err
-	}
-
-	ackStream, err := c.store.Ack(context.Background())
-	if err != nil {
-		subscribeStream.CloseSend()
-		return nil, err
-	}
+func (s *subscribe) Listen(handler func(ctx context.Context, msgs ...Message) error) error {
+	s.handler = handler
+	go func() {
+		barrier := make(chan struct{}, s.options.parallelism)
+		for idx := 0; idx < s.options.parallelism; idx++ {
+			barrier <- struct{}{}
+		}
+		for {
+			select {
+			case <-s.closeC:
+				// TODO log
+				goto CLOSE
+			case msg := <-s.messageC:
+				if msg == nil {
+					goto CLOSE // TODO(wenfeng) how to process if msg is nil?
+				}
+				<-barrier
+				l := len(s.messageC)
+				var msgs []Message
+				msgs = append(msgs, msg)
+				for idx := 0; idx < l && idx < s.options.batchSize; idx++ {
+					msgs = append(msgs, <-s.messageC)
+				}
+				go func() {
+					// TODO(wenfeng) How to process error?
+					_ = s.handler(context.Background(), msgs...)
+					barrier <- struct{}{}
+				}()
+			}
+		}
+	CLOSE:
+		close(barrier)
+	}()
+	return s.startReceive()
+}
 
-	ch := make(chan Message, 32)
-	return &subscribe{
-		store:           c.store,
-		options:         opts,
-		subscribeStream: subscribeStream,
-		ackStream:       ackStream,
-		messagec:        ch,
-		state:           stateInitialized,
-	}, nil
+func (s *subscribe) Close() error {
+	s.closeOnce.Do(func() {
+		_ = s.subscribeStream.CloseSend()
+		s.subscribeStream = nil
+		_ = s.ackStream.CloseSend()
+		s.ackStream = nil
+		close(s.messageC)
+		close(s.closeC)
+		s.state = stateClosed
+	})
+	return nil
 }
 
-func (s *subscribe) release() {
-	s.subscribeStream.CloseSend()
-	s.subscribeStream = nil
-	s.ackStream.CloseSend()
-	s.ackStream = nil
-	close(s.messagec)
-	s.state = stateClosed
+func newSubscriber(cc *grpc.ClientConn, opts *subscribeOptions) Subscriber {
+	return &subscribe{
+		store:    proxypb.NewStoreProxyClient(cc),
+		options:  opts,
+		messageC: make(chan Message, 32),
+		closeC:   make(chan struct{}),
+		state:    stateInitialized,
+	}
 }
 
 func (s *subscribe) SubscriptionID() string {
-	return s.options.SubscriptionID
+	return s.options.subscriptionID
 }
 
-func (s *subscribe) Subscribe(ctx context.Context) (<-chan Message, error) {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-	if s.state == stateRunning {
-		return s.messagec, nil
+func (s *subscribe) Send(ctx context.Context, event *cloudevents.BatchEvent) (*emptypb.Empty, error) {
```

Comment:
What is the use scenario of this method？

Reply:
see example: https://github.com/linkall-labs/vanus/pull/413/files#diff-1b1b5c22b325b98c8633760d2c12138cdcf4e5cdf0d76a3898cf7c50448ef3aeR149.

SDK both support Active and Inactive modes:

- Active mode: SDK builds a gRPC stream within Gateway to receive events
- 
- Inactive mode: default mode, in this mode, SDK starts a server to listen port, and Subscription can push events directly to it.